### PR TITLE
FIX: Fixes #138 For Fluent installed in existing project

### LINKED ISSUE #138: Default translated field oddities
@tractorcow This issue relates to #132 with related bug(s). I have not verified if the following were introduced recently or existed prior to #132 being fixed:

## Env

* Fluent: dev-master as 3.2.2 with aliases enabled in YML for all locales and FluentExtension + FluentFilteredExtension enabled for SiteTree
* Newly updated, raw SS 3.1.12 (No project-code whatsoever)
* PHP 5.5 / Ubuntu 14.04
* Chrome 44 (CMS) Firefox 39 (Frontend)

## Bug reproduction

* [OK] Login to the CMS
* [OK] Select "English (New Zealand)" from LHS locale menu
* [OK] Create new top-level "Page"
* [OK] Enter a title in English; e.g. "TEST 01"
* [OK] Press "Save & Publish"
* [OK] Observe "Page name" field reads "TEST 01" and "Navigation label" reads" "TEST 01"
* [OK] Navigate to /en/test in the frontend, observe Menu title reads "TEST 01" and page-heading reads "TEST 01"
* [OK] Select "Arabic (Egypt)" from LHS locale menu
* [FAIL] Observe "Page name" suddenly displays "New Page" (Could be session related as this seems inconsistent..)
* [OK] Enter a "Page name" in Arabic; e.g. "TEST 01 Arabic"
* [FAIL] Focusing off the "Page name" field does _not_ auto-change the "Navigation label" field as it would normally do (No JS errors observed) - this could be related to the above as it also seems inconsistent)
* [OK] Press "Save & Publish"
* [FAIL] Observe "Page name" field reads "TEST 01" (Expected "TEST 01 Arabic")
* [FAIL] Navigate to /en/test in the frontend, observe Menu title reads "TEST 01 Arabic" and page-heading reads "TEST 01" (Expected both to read "TEST 01")
* [FAIL] Navigate to /ar/test in the frontend, observe Menu title reads "TEST 01 Arabic" and Page-heading reads "TEST 01" (Expected both to read "TEST 01 Arabic")

Relating to #132 I see similar behaviour for _pre-existing_ pages:

* [OK] Login to the CMS
* [OK] Select "English (New Zealand)" from LHS locale menu
* [OK] Open an existing default page e.g "About us" (ID #2)
* [OK] Select "Arabic (Egypt)" from LHS locale menu
* [OK] Enter a new "Page name" in Arabic; e.g. "About us in Arabic"
* [OK] Press "Save & Publish"
* [FAIL] Observe the "Page name" fiel din thre CMS reads "About us" but the "Nvaigatioin field" reads "About us in Arabic"
* [FAIL] Navigate to /en/about-us in the frontend, observe Menu title reads "About us in Arabic" and page-heading reads "About us" (Expected both to read "About us")
* [FAIL] Navigate to /ar/about-us in the frontend, observe Menu title reads "About us in Arabic" and page-heading reads "About us" (Expected both to read "About us in Arabic")

### PR DESCRIPTION
**Pull request by phptek**
By default the CMS will update non-Locale aware fields, regardless of
selected CMS locale, with the selected Locale's data.

This patch therefore:

- Assumes non-localised field-data == default_locale aware field-data
- Caches original non-localised field-data, and re-applies it

### CODE REVIEW

**Comment by phptek** [comment]
Please note: I have run this through the first two scenarios posted in #138, and apart from the `MenuTitle` "anomaly" which I have hacked around using JS in my own project, this passes. Obviously you'll want to run through it and comment accordingly - have at it :-)

**Comment by tractorcow** [comment]
By non-locale aware, you mean fields that aren't created with Field_locale variants in the database?

**Comment by tractorcow** [comment]
Could you please include a unit test that captures the issue as simply as possible?

If you want to set a custom DB state, it's fine if you use low level SQL queries for your setup instead of the unit test fixtures.

**Comment by tractorcow** [comment]
Am I correct in assuming that this is the test script I should use to test this fix?

```
[OK] Login to CMS
[OK] In CMS Select "English (New Zealand)" locale
[OK] Create new page with title "TEST 01"
[OK] Observe navigation label changes to "TEST 01"
[OK] Observe URL Segment changes to /en/test-01
[OK] Save & Publish
[OK] View in the frontend: as /en/test-01
[OK] Observe Menu title reads "Test 01"
[OK] Observe Page title reads "TEST 01"
[OK] Observe URL segment reads "/en/test-01"
[OK] In CMS Select "Chinese (Taiwan)" locale
[OK] Modify page title to read "TEST 01 Something in Taiwanese"
[FAIL] Observe navigation label changes to "TEST 01 Something in Taiwanese"
[OK] Observe URL Segment is still /en/test-01
[OK] Save & Publish
[OK] View in the frontend: as /en/test-01
[FAIL] Observe Menu title still reads "Test 01" (It reads "TEST 01 Something in Taiwanese")
[OK] Observe Page title reads "TEST 01"
[OK] Observe URL segment reads "/en/test-01"
[OK] View in the frontend: as /tw/test-01
[FAIL] Observe Menu title still reads "Test 01" (It reads "TEST 01 Something in Taiwanese")
[OK] Observe Page title reads "TEST 01 Something in Taiwanese"
[OK] Observe URL segment reads "/en/test-01"
[OK] In CMS Re-select "English (New Zealand)" locale
[OK] Change the Menu Title back to "TEST 01"
[OK] Save & Publish
[OK] View in the frontend: as /en/test-01
[OK] Observe Menu title reads "Test 01"
[OK] View in the frontend: as /tw/test-01
[OK] Observe Menu title still reads "Test 01"
```

File: code/extensions/FluentExtension.php [hunk 1]
```diff
@@ -16,6 +16,16 @@ class FluentExtension extends DataExtension {
 	 * @var boolean
 	 */
 	protected static $disable_fluent_fields = false;
+    
+    /**
+	 * Caches the original DB field-values of all non-Locale, YML-configured fields 
+	 * during onBeforeWrite() so their original values can be written-back to them. 
+	 * If we don't do this, the CMS interprets POSTed form-fields as _the_ data to write to 
+	 * these non Fluent-enabled fields, regardless of the currently selected locale.
```
Comment (tractorcow) [inline code/extensions/FluentExtension.php]: Does this issue only affect MenuTitle, or do other fields have this issue also?

**Comment by tractorcow** [comment]
If you don't want to write a unit test, it's probably enough if you just attach a manual testing script.

As I am prone to do, I'll probably write another fix for this that works around the fragile nature of the translation process. :)

**Comment by phptek** [comment]
>By non-locale aware, you mean fields that aren't created with Field_locale variants in the database?
Yes. I want to call them "Default" but we'll get confused with "default_locale"

>Could you please include a unit test that captures the issue as simply as possible?
I can of course - but as I'm under time pressure (aren't we always?) if you're likely to re-write or refactor the work, then it may not be the best use of my time.

>Am I correct in assuming that this is the test script I should use to test this fix?
No. Ignore "MenuTitle" related issues, like I said, I have hacked around this issue in my own project's JS. We should raise this as a new issue.

File: code/extensions/FluentExtension.php (hunk 1 above)
Comment (phptek) [inline code/extensions/FluentExtension.php]: See my most recent comment below in #141

**Comment by tractorcow** [comment]
Totally understand the time pressure; I'm in a similar situation, but I do want to help!

So, to basically boil down the issue:

- Save a page in default locale (setting both translated and un-translated field)
- Go to another locale, and override BOTH fields with another value
- Go back to the default locale, and expect to see both values as they were originally set, but actually only the translated field is correct, and the untranslated (or non-locale aware) field is set to the value you put in the other locale?

**Comment by tractorcow** [comment]
> By default the CMS will update non-Locale aware fields, regardless of
selected CMS locale, with the selected Locale's data.

I'm not 100% clear on your use case here, but I'm probably about to say "this is the way the module is meant to work". :) If a field isn't localised, it shouldn't matter what locale you are in; It saves the value you enter.

I've looked at #138 but the first two scripts you mention both have MenuTitle in them, so I'm having trouble mentally filtering it out. Maybe (for my poor brain's sake) you could re-transcribe them and post them below as it pertains to this issue?

**Comment by phptek** [comment]
Thanks @tractorcow, I'll be online later to help you...umm help me.

**Comment by phptek** [comment]
@tractorcow here you go, the simplest reproduction steps I can do, stripped of anything relating to `MenuTitle` :-)

* Assumes a pre-existing SS3.1 project, originally created without Fluent. 
* Fluent is subsequently installed on top of pre-existing content.
* mysite/_config/config.yml features `FluentFilteredExtension` applied to `SiteTree` and a `default_locale` of `en_NZ`.
* Taken from Fluent `master` at `5856da`

PRE-EXISTING PAGE (NEW PAGES ARE A-OK)

[OK] Open an existing default page e.g "About us" (ID #2)
[OK] Select "Arabic (Egypt)" from LHS locale menu
[OK] Modify the "Page name" in Arabic; e.g. "About us in Arabic"
[OK] Press "Save & Publish"	
[FAIL] Navigate to /en/about-us in the frontend, observe page-title reads "About us in Arabic"
[OK] Navigate to /ar/about-us in the frontend, observe page-title reads "About us in Arabic"

* Now run with #138 `43a2c8ce53b` and repeat the above, all seems to work A-OK.

**Comment by tractorcow** [comment]
Thanks for that @phptek ; I'll try to check that out soon and get back to you. :)

**Comment by tractorcow** [comment]
Yeah, I think the "Fluent is subsequently installed on top of pre-existing content." case is what I wasn't testing properly.

Potential fix: When a record is loaded (not when it's saved) make sure that the Field_default_locale is populated by the value if Field, if it isn't set.

**Comment by phptek** [comment]
@tractorcow I'm running with a module fork which I'd rather not do if I can avoid it. What do you think the solution should be for this?

**Comment by tractorcow** [comment]
Sorry @phptek I haven't had much time to work on fluent this week. I'll see if I can spend some time on it tomorrow.

**Comment by tractorcow** [comment]
Fixed at #147 . I fixed it on loading of the record, however, (rather than on save).

Does this solve your problem @phptek ?

**Comment by tractorcow** [comment]
Ah, merging that automatically closed this issue. Is that ok? Do you need to re-open this to confirm it's fixed @phptek ?

**Comment by phptek** [comment]
@tractorcow at last got back onto this - thanks: Just confirming that dev-master at `1b995c15c` seems to resolve the issue using the repro steps above.

Closing...

**Comment by tractorcow** [comment]
Yay thanks. I know that feel... github issues can get hard to stay on top of. :)

**Pull request closed without merging**